{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Incorporating Phage Biomass\n",
    "This notebook is adapted from:\n",
    "\n",
    "Hadrien Delattre, Kalesh Sasidharan, Orkun S Soyer.\n",
    "Life Science Alliance Nov 2020, 4 (1). DOI: 10.26508/lsa.202000869\n",
    "\n",
    "https://github.com/OSS-Lab/FBAhv/tree/master"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [],
   "source": [
    "import cobra\n",
    "from cobra import Reaction\n",
    "from collections import Counter\n",
    "from Bio import SeqIO\n",
    "\n",
    "from cobra.io import read_sbml_model, write_sbml_model"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Constants and Conventions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Metabolite (verbose to model) definitions\n",
    "MET_IDS = {\n",
    "    \"atp\": \"ATP[c]\",\n",
    "    \"ctp\": \"CTP[c]\",\n",
    "    \"gtp\": \"GTP[c]\",\n",
    "    \"utp\": \"UTP[c]\",\n",
    "    \"amp\": \"AMP[c]\",\n",
    "    \"cmp\": \"CMP[c]\",\n",
    "    \"gmp\": \"GMP[c]\",\n",
    "    \"ump\": \"UMP[c]\",\n",
    "    \"datp\": \"dATP[c]\",\n",
    "    \"dctp\": \"dCTP[c]\",\n",
    "    \"dgtp\": \"dGTP[c]\",\n",
    "    \"dttp\": \"dTTP[c]\",\n",
    "    \"damp\": \"dAMP[c]\",\n",
    "    \"dcmp\": \"dCMP[c]\",\n",
    "    \"dgmp\": \"dGMP[c]\",\n",
    "    \"dtmp\": \"dTMP[c]\",\n",
    "    \"A\": \"L_Alanine[c]\",\n",
    "    \"R\": \"L_Arginine[c]\",\n",
    "    \"N\": \"L_Asparagine[c]\",\n",
    "    \"D\": \"L_Aspartate[c]\",\n",
    "    \"C\": \"L_Cysteine[c]\",\n",
    "    \"Q\": \"L_Glutamine[c]\",\n",
    "    \"E\": \"L_Glutamate[c]\",\n",
    "    \"G\": \"Glycine[c]\",\n",
    "    \"H\": \"L_Histidine[c]\",\n",
    "    \"I\": \"L_Isoleucine[c]\",\n",
    "    \"L\": \"L_Leucine[c]\",\n",
    "    \"K\": \"L_Lysine[c]\",\n",
    "    \"M\": \"L_Methionine[c]\",\n",
    "    \"F\": \"L_Phenylalanine[c]\",\n",
    "    \"P\": \"L_Proline[c]\",\n",
    "    \"S\": \"L_Serine[c]\",\n",
    "    \"T\": \"L_Threonine[c]\",\n",
    "    \"W\": \"L_Tryptophan[c]\",\n",
    "    \"Y\": \"L_Tyrosine[c]\",\n",
    "    \"V\": \"L_Valine[c]\",\n",
    "    \"h2o\": \"H2O[c]\",\n",
    "    \"adp\": \"ADP[c]\",\n",
    "    \"Pi\": \"Orthophosphate[c]\",\n",
    "    \"h\": \"H[c]\",\n",
    "    \"PPi\": \"Diphosphate[c]\",  # Pyrophosphate\n",
    "}\n",
    "\n",
    "# Nucleotides Dictionary in g/mol\n",
    "NUCLEOTIDE_MASS = {\n",
    "    \"atp\": 507.181,\n",
    "    \"gtp\": 483.15644,\n",
    "    \"ctp\": 523.18062,\n",
    "    \"utp\": 484.14116,\n",
    "    \"datp\": 491.1816,\n",
    "    \"dgtp\": 467.1569,\n",
    "    \"dctp\": 482.1683,\n",
    "    \"dttp\": 507.181,\n",
    "    \"damp\": 331.068,\n",
    "    \"dgmp\": 347.063,\n",
    "    \"dcmp\": 307.057,\n",
    "    \"dtmp\": 322.057,\n",
    "}\n",
    "# Amino Acids Dictionary g/mol\n",
    "AMINO_ACID_MASS = {\n",
    "    \"A\": 89.09322,  # Alaline\n",
    "    \"R\": 174.201,  # Arginine\n",
    "    \"N\": 132.118,  # Asparagine\n",
    "    \"D\": 133.1027,  # Aspartate\n",
    "    \"C\": 121.158,  # Cysteine\n",
    "    \"Q\": 146.14458,  # Glutamine\n",
    "    \"E\": 147.1293,  # Glutamate\n",
    "    \"G\": 75.06664,  # Glycine\n",
    "    \"H\": 155.15468,  # Histidine\n",
    "    \"I\": 131.17296,  # Isoleucine\n",
    "    \"L\": 131.17296,  # Leucine\n",
    "    \"K\": 146.18764,  # Lysine\n",
    "    \"M\": 149.21238,  # Methionine\n",
    "    \"F\": 165.18918,  # Phenylalanine\n",
    "    \"P\": 115.1305,  # Proline\n",
    "    \"S\": 105.09262,  # Serine\n",
    "    \"T\": 119.1192,  # Threonine\n",
    "    \"W\": 204.22526,  # Tryptophan\n",
    "    \"Y\": 181.18858,  # Tyrosine\n",
    "    \"V\": 117.14638,  # Valine\n",
    "}\n",
    "# Misc. Dictionary\n",
    "MISC_MASS = {\n",
    "    \"PPi\": 173.94332,  # Pyrophosphate\n",
    "}\n",
    "# Avogadro's Number\n",
    "N_A = 6.0221409e23\n",
    "# ATP requirement coefficients\n",
    "# source: Queka, Dietmaira, Hanschob, Martíneza, Borthb, Nielsen\n",
    "# Journal of Biotechnology 184 (2014) 172–178\n",
    "K_ATP_PROTEIN = 4.3\n",
    "K_ATP_DNA = 1.4\n",
    "K_ATP_RNA = 0.4\n",
    "\n",
    "PROTEINS_PER_MRNA = 540\n",
    "\n",
    "K_PPI = 1  # pyrophosphate production ratio\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Defining Virus Protein Composition"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [],
   "source": [
    "# NOTE: the keys in this dictionary correspond to the id of the product in the genbank file\n",
    "VIRUS_COMPOSITION = {\n",
    "    \"PHM2\": {  # P-HM2                     [Cg]\n",
    "        \"Cg\": 1,\n",
    "        \"proteins\": {\n",
    "            \"ADO99900.1\": 985,\n",
    "            \"ADO99896.1\": 12,\n",
    "            \"ADO99878.1\": 870,\n",
    "            \"ADO99884.1\": 155,\n",
    "            \"ADO99893.1\": 5,\n",
    "            \"ADO99885.1\": 12,\n",
    "            \"ADO99886.1\": 6,\n",
    "            \"ADO99864.1\": 24,\n",
    "            \"ADO99942.1\": 18,\n",
    "            \"ADO99865.1\": 12,\n",
    "            \"ADO99863.1\": 12,\n",
    "            \"ADO99793.1\": 6,\n",
    "            \"ADO99862.1\": 6,\n",
    "            \"ADO99803.1\": 3,\n",
    "            \"ADO99913.1\": 3,\n",
    "            \"ADO99860.1\": 1,\n",
    "            \"ADO99798.1\": 1,\n",
    "            \"ADO99895.1\": 145,\n",
    "            \"ADO99800.1\": 1,\n",
    "            \"ADO99799.1\": 1,\n",
    "            \"ADO99874.1\": 40,\n",
    "            \"ADO99872.1\": 21,\n",
    "            \"ADO99794.1\": 6,\n",
    "            \"ADO99894.1\": 138,\n",
    "            \"ADO99903.1\": 6,\n",
    "            \"ADO99887.1\": 6,\n",
    "            \"ADO99947.1\": 3,\n",
    "            \"ADO99993.1\": 2,\n",
    "            \"ADO99924.1\": 1,\n",
    "            \"ADO99795.1\": 6,\n",
    "            \"ADO99796.1\": 6,\n",
    "        },\n",
    "    },\n",
    "}\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Utility Functions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [],
   "source": [
    "def multiply_counter(counter, n):\n",
    "    \"\"\"return a collection.Counter whose values have been multiplied by a\n",
    "    scalar\"\"\"\n",
    "    # the multiplicand does not need to be a collections.Counter, but at least\n",
    "    # it needs to be a dict\n",
    "    assert isinstance(counter, dict)\n",
    "    assert float(n)\n",
    "    return Counter({key: value * n for key, value in counter.items()})\n",
    "\n",
    "\n",
    "def get_virus_names(virus_record):\n",
    "    \"\"\"Return a tuple containing a short name and a\n",
    "    full name for the virus. This is an alternative to how it is done in Sean's\n",
    "    genVBOF.\"\"\"\n",
    "    long_name = virus_record.description.rstrip(\", complete genome\")\n",
    "    # try to get a short name for the virus by taking the prefix of the first CDS\n",
    "    short_name = (\n",
    "        next(feature for feature in virus_record.features if feature.type == \"CDS\")\n",
    "        .qualifiers[\"locus_tag\"][0]\n",
    "        .rstrip(\"_gp01\")\n",
    "    )\n",
    "    return (short_name, long_name)\n",
    "\n",
    "\n",
    "def reverse_complement(seq):\n",
    "    rc = \"\"\n",
    "    for nuc in seq[::-1]:\n",
    "        if nuc == \"A\":\n",
    "            rc += \"T\"\n",
    "        elif nuc == \"T\":\n",
    "            rc += \"A\"\n",
    "        elif nuc == \"C\":\n",
    "            rc += \"G\"\n",
    "        elif nuc == \"G\":\n",
    "            rc += \"C\"\n",
    "        else:\n",
    "            raise ValueError(\"Invalid nucleotide: %s\" % nuc)\n",
    "    return rc"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Main Function"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [],
   "source": [
    "def generate_virus_biomass(virus_record, model, model_name=None):\n",
    "    \"\"\"New version of the genVBOF function by Hadrien.\n",
    "    Builds a Virus Biomass Objective Function (basically a virus biomass\n",
    "    production reaction, from aminoacids and nucleotides) from a genbank (.gb)\n",
    "    file. Modified from the original by Jordan C. Rozum and William Sineath.\n",
    "\n",
    "    Params:\n",
    "    - virus_record: genbank record of a virus (output from Bio.SeqIO.parse)\n",
    "    - model: a cobra metabolic model (cobra.core.model.Model)\n",
    "\n",
    "    Returns:\n",
    "    - virus biomass objective function (cobra.core.reaction.Reaction)\n",
    "    \"\"\"\n",
    "\n",
    "    # VIRUS IDENTIFICATION\n",
    "    taxonomy = \" \".join(\n",
    "        [taxon.lower() for taxon in virus_record.annotations[\"taxonomy\"]]\n",
    "    )\n",
    "    if \"ssp7\" not in taxonomy and \"hm2\" not in taxonomy:\n",
    "        raise NotImplementedError(\n",
    "            \"Virus family is not supported: Unable to create VBOF. Consult _README\"\n",
    "        )\n",
    "    short_name, full_name = get_virus_names(virus_record)\n",
    "\n",
    "    # AMINOACID COUNT\n",
    "    all_cds = [feature for feature in virus_record.features if feature.type == \"CDS\"]\n",
    "    # Check that our own VIRUS_COMPOSITION dict contain exactly the\n",
    "    # proteins defined in the genbank file, no more, no less.\n",
    "\n",
    "    protein_ids_in_gb_file = {\n",
    "        cds.qualifiers[\"protein_id\"][0] for cds in all_cds\n",
    "    }  # not allowing set to be created\n",
    "    protein_ids_in_our_data = {\n",
    "        protein_id for protein_id in VIRUS_COMPOSITION[short_name][\"proteins\"]\n",
    "    }\n",
    "    assert protein_ids_in_our_data.issubset(protein_ids_in_gb_file)\n",
    "    # make sure that gb file has all the protein ids in our data\n",
    "\n",
    "    virus_aa_composition = Counter()\n",
    "    virus_mrna_composition = Counter()\n",
    "    # protein name -> number of atp involved in its peptide bonds formations\n",
    "    # (accounting for the number of copies of protein)\n",
    "    peptide_bond_formation = dict()\n",
    "    for cds in all_cds:\n",
    "        if cds.qualifiers[\"protein_id\"][0] in protein_ids_in_our_data:\n",
    "            protein_id = cds.qualifiers[\"protein_id\"][0]\n",
    "            aa_sequence = cds.qualifiers[\"translation\"][0]\n",
    "            dna_sequence = virus_record.seq[cds.location.start : cds.location.end]\n",
    "            if cds.location.strand == -1:\n",
    "                dna_sequence = reverse_complement(dna_sequence)\n",
    "            aa_count = Counter(aa_sequence)\n",
    "            nc_count = Counter(dna_sequence)\n",
    "            copies_per_virus = VIRUS_COMPOSITION[short_name][\"proteins\"][\n",
    "                protein_id\n",
    "            ]  ##can change logic to count\n",
    "            virus_aa_composition += multiply_counter(aa_count, copies_per_virus)\n",
    "            virus_mrna_composition += multiply_counter(nc_count, copies_per_virus)\n",
    "\n",
    "            peptide_bond_formation[protein_id] = (\n",
    "                len(aa_sequence) * K_ATP_PROTEIN - K_ATP_PROTEIN\n",
    "            ) * copies_per_virus\n",
    "\n",
    "    # [3] Precursor frequency\n",
    "    # Genome                            [Nucleotides]\n",
    "    Cg = VIRUS_COMPOSITION[short_name][\"Cg\"]  # number of genome copies per virus\n",
    "    virus_nucl_count = Counter(str(virus_record.seq))\n",
    "    countA = virus_nucl_count[\"A\"]\n",
    "    countC = virus_nucl_count[\"C\"]\n",
    "    countG = virus_nucl_count[\"G\"]\n",
    "    countT = virus_nucl_count[\"T\"]\n",
    "    antiA = countT\n",
    "    antiC = countG\n",
    "    antiG = countC\n",
    "    antiT = countA\n",
    "\n",
    "    # Note: mRNA is 1-stranded, so we don't track the \"anti\" parts\n",
    "    # Also: we didn't bother to replace T->U in the string; we do it here\n",
    "    countA_rna = virus_mrna_composition[\"A\"] / PROTEINS_PER_MRNA\n",
    "    countC_rna = virus_mrna_composition[\"C\"] / PROTEINS_PER_MRNA\n",
    "    countG_rna = virus_mrna_composition[\"G\"] / PROTEINS_PER_MRNA\n",
    "    countU_rna = virus_mrna_composition[\"T\"] / PROTEINS_PER_MRNA\n",
    "\n",
    "    # Count summation (not used)\n",
    "    # totNTPS = Cg * (countA + countC + countG + countT + antiA + antiC + antiG + antiT)\n",
    "    # totAA = sum(count for count in virus_aa_composition.values())\n",
    "\n",
    "    # [4] VBOF Calculations\n",
    "    # Nucleotides\n",
    "    # mol.dntps/mol.virus\n",
    "    V_a = Cg * (countA + antiA)\n",
    "    V_c = Cg * (countC + antiC)\n",
    "    V_g = Cg * (countG + antiG)\n",
    "    V_t = Cg * (countT + antiT)\n",
    "    # g.dnmps/mol.virus\n",
    "    G_a = V_a * NUCLEOTIDE_MASS[\"damp\"]\n",
    "    G_c = V_c * NUCLEOTIDE_MASS[\"dcmp\"]\n",
    "    G_g = V_g * NUCLEOTIDE_MASS[\"dgmp\"]\n",
    "    G_t = V_t * NUCLEOTIDE_MASS[\"dtmp\"]\n",
    "\n",
    "    # mol.ntps (mrna) / mol.virus\n",
    "    V_a_rna = countA_rna\n",
    "    V_c_rna = countC_rna\n",
    "    V_g_rna = countG_rna\n",
    "    V_u_rna = countU_rna\n",
    "\n",
    "    # Amino Acids\n",
    "    # g.a/mol.virus\n",
    "    G_aa = {\n",
    "        aa: count * AMINO_ACID_MASS[aa] for aa, count in virus_aa_composition.items()\n",
    "    }\n",
    "    # Molar mass of the virus particles (g.virus/mol.virus); note that the mRNA is not packaged\n",
    "    M_v = (G_a + G_c + G_g + G_t) + sum(G_aa.values())\n",
    "\n",
    "    # Stoichiometric coefficients\n",
    "    # Nucleotides [mmol.ntps/g.virus] (for the genome)\n",
    "    S_datp = 1000 * (V_a / M_v)\n",
    "    S_dctp = 1000 * (V_c / M_v)\n",
    "    S_dgtp = 1000 * (V_g / M_v)\n",
    "    S_dttp = 1000 * (V_t / M_v)\n",
    "\n",
    "    # Amino acids [mmol.aa/g.virus]\n",
    "    S_aa = {aa: 1000 * V_aa / M_v for aa, V_aa in virus_aa_composition.items()}\n",
    "\n",
    "    # mRNA is handeled differently because the virus doesn't pack it. We will\n",
    "    # allow it to decay into NMPs at a rate governed by the protein to mrna\n",
    "    # ratio (for mass conservation)\n",
    "    S_atp_rna = 1000 * (V_a_rna / M_v)\n",
    "    S_ctp_rna = 1000 * (V_c_rna / M_v)\n",
    "    S_gtp_rna = 1000 * (V_g_rna / M_v)\n",
    "    S_utp_rna = 1000 * (V_u_rna / M_v)\n",
    "    S_amp_rna = 1000 * (V_a_rna / M_v)\n",
    "    S_cmp_rna = 1000 * (V_c_rna / M_v)\n",
    "    S_gmp_rna = 1000 * (V_g_rna / M_v)\n",
    "    S_ump_rna = 1000 * (V_u_rna / M_v)\n",
    "\n",
    "    # Energy requirements -(2+n_proteins/PROTEINS_PER_MRNA) is coming from total\n",
    "    # strands: 2 for DNA, 1 for mRNA This is because we are counting the number\n",
    "    # of bonds (one less than number of nucleotides in a strand). Not that this\n",
    "    # term will be exceedingly small compared to the rest, so accuracy here is\n",
    "    # not critical.\n",
    "    n_proteins = len(VIRUS_COMPOSITION[short_name][\"proteins\"])\n",
    "    S_ppi = (\n",
    "        S_atp_rna\n",
    "        + S_ctp_rna\n",
    "        + S_gtp_rna\n",
    "        + S_utp_rna\n",
    "        + S_datp\n",
    "        + S_dctp\n",
    "        + S_dgtp\n",
    "        + S_dttp\n",
    "        - (2 + n_proteins / PROTEINS_PER_MRNA) * 1000 / M_v\n",
    "    ) * K_PPI\n",
    "\n",
    "    # Proteome: Peptide bond formation [ATP + H2O]\n",
    "    # Note: ATP used in this process is denoated as ATPe/Ae [e = energy version]\n",
    "    V_Ae = sum(peptide_bond_formation.values())\n",
    "    S_Ae = 1000 * (V_Ae / M_v)\n",
    "\n",
    "    # [5] VBOF Reaction formatting and output\n",
    "    # Left-hand terms: Nucleotides\n",
    "    S_ATP = S_Ae * -1 - S_atp_rna\n",
    "    S_CTP = S_ctp_rna * -1\n",
    "    S_GTP = S_gtp_rna * -1\n",
    "    S_UTP = S_utp_rna * -1\n",
    "    S_AMP = S_amp_rna\n",
    "    S_CMP = S_cmp_rna\n",
    "    S_GMP = S_gmp_rna\n",
    "    S_UMP = S_ump_rna\n",
    "    S_DATP = S_datp * -1\n",
    "    S_DCTP = S_dctp * -1\n",
    "    S_DGTP = S_dgtp * -1\n",
    "    S_DTTP = S_dttp * -1\n",
    "\n",
    "    # Left-hand terms: Amino Acids\n",
    "    S_AAf = {aa: -coef for aa, coef in S_aa.items()}\n",
    "    # Left-hand terms: Energy Requirements\n",
    "    S_H2O = S_Ae * -1\n",
    "    # Right-hand terms: Energy Requirements\n",
    "    S_ADP = S_Ae\n",
    "    S_Pi = S_Ae\n",
    "    S_H = S_Ae\n",
    "\n",
    "    reaction_name = short_name + \"_prodrxn_VN\"\n",
    "    virus_reaction = Reaction(reaction_name)\n",
    "    virus_reaction.name = full_name + \" production reaction\"\n",
    "    virus_reaction.subsystem = \"Virus Production\"\n",
    "    virus_reaction.lower_bound = 0\n",
    "    virus_reaction.upper_bound = 1000\n",
    "    model.add_reactions([virus_reaction])\n",
    "    virus_reaction.add_metabolites(\n",
    "        (\n",
    "            {\n",
    "                MET_IDS[\"atp\"]: S_ATP,\n",
    "                MET_IDS[\"ctp\"]: S_CTP,\n",
    "                MET_IDS[\"gtp\"]: S_GTP,\n",
    "                MET_IDS[\"utp\"]: S_UTP,\n",
    "                MET_IDS[\"amp\"]: S_AMP,\n",
    "                MET_IDS[\"cmp\"]: S_CMP,\n",
    "                MET_IDS[\"gmp\"]: S_GMP,\n",
    "                MET_IDS[\"ump\"]: S_UMP,\n",
    "                MET_IDS[\"datp\"]: S_DATP,\n",
    "                MET_IDS[\"dctp\"]: S_DCTP,\n",
    "                MET_IDS[\"dgtp\"]: S_DGTP,\n",
    "                MET_IDS[\"dttp\"]: S_DTTP,\n",
    "                MET_IDS[\"A\"]: S_AAf[\"A\"],\n",
    "                MET_IDS[\"R\"]: S_AAf[\"R\"],\n",
    "                MET_IDS[\"N\"]: S_AAf[\"N\"],\n",
    "                MET_IDS[\"D\"]: S_AAf[\"D\"],\n",
    "                MET_IDS[\"C\"]: S_AAf[\"C\"],\n",
    "                MET_IDS[\"Q\"]: S_AAf[\"Q\"],\n",
    "                MET_IDS[\"E\"]: S_AAf[\"E\"],\n",
    "                MET_IDS[\"G\"]: S_AAf[\"G\"],\n",
    "                MET_IDS[\"H\"]: S_AAf[\"H\"],\n",
    "                MET_IDS[\"I\"]: S_AAf[\"I\"],\n",
    "                MET_IDS[\"L\"]: S_AAf[\"L\"],\n",
    "                MET_IDS[\"K\"]: S_AAf[\"K\"],\n",
    "                MET_IDS[\"M\"]: S_AAf[\"M\"],\n",
    "                MET_IDS[\"F\"]: S_AAf[\"F\"],\n",
    "                MET_IDS[\"P\"]: S_AAf[\"P\"],\n",
    "                MET_IDS[\"S\"]: S_AAf[\"S\"],\n",
    "                MET_IDS[\"T\"]: S_AAf[\"T\"],\n",
    "                MET_IDS[\"W\"]: S_AAf[\"W\"],\n",
    "                MET_IDS[\"Y\"]: S_AAf[\"Y\"],\n",
    "                MET_IDS[\"V\"]: S_AAf[\"V\"],\n",
    "                MET_IDS[\"h2o\"]: S_H2O,\n",
    "                MET_IDS[\"adp\"]: S_ADP,\n",
    "                MET_IDS[\"Pi\"]: S_Pi,\n",
    "                MET_IDS[\"h\"]: S_H,\n",
    "                MET_IDS[\"PPi\"]: S_ppi,\n",
    "            }\n",
    "        )\n",
    "    )\n",
    "    return virus_reaction"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Initial modifications to the model\n",
    "\n",
    "Here, alter the DNA and RNA formation reactions so that they are built from (d)NTP and produce pyrophosphate.\n",
    "\n",
    "We print a before and after comparison of the optimal biomass. Note that it does not significantly change."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "BEFORE:\n",
      "BIOMASS = 0.09846248800237327\n",
      "AFTER:\n",
      "BIOMASS = 0.09846248800237146\n"
     ]
    }
   ],
   "source": [
    "virus_record = SeqIO.read(\"./model_files/hm2.gbff\", \"genbank\")\n",
    "model = read_sbml_model(\"./model_files/ProchlorococcusMED4v1.xml\")\n",
    "\n",
    "\n",
    "print(\"BEFORE:\")\n",
    "print(f\"BIOMASS = {model.slim_optimize()}\")\n",
    "\n",
    "m1 = model.reactions.BDNA.metabolites\n",
    "nt = 0\n",
    "for m, s in m1.items():\n",
    "    if m.id in [\"dATP[c]\", \"dCTP[c]\", \"dGTP[c]\", \"dTTP[c]\"]:\n",
    "        nt += s\n",
    "# Note: we are neglecting the number of strands, but this will have a neglible\n",
    "# effect since this will be much smaller than the number of nucleotides\n",
    "\n",
    "m1[model.metabolites.get_by_id(\"Diphosphate[c]\")] = -nt\n",
    "model.reactions.BDNA.add_metabolites(m1)\n",
    "\n",
    "m2 = {}\n",
    "nt = 0\n",
    "original = dict(model.reactions.BRNA.metabolites.items())\n",
    "for m, s in model.reactions.BRNA.metabolites.items():\n",
    "    if m.id in [\"AMP[c]\", \"CMP[c]\", \"GMP[c]\", \"UMP[c]\"]:\n",
    "        nt += s\n",
    "        ntp = m.id.replace(\"M\", \"T\")\n",
    "        if ntp == \"ATP[c]\":\n",
    "            met = model.metabolites.get_by_id(ntp)\n",
    "            m2[met] = s + model.reactions.BRNA.metabolites[met]\n",
    "        else:\n",
    "            m2[model.metabolites.get_by_id(ntp)] = s\n",
    "    elif m.id != \"ATP[c]\":\n",
    "        m2[model.metabolites.get_by_id(m.id)] = s\n",
    "\n",
    "m2[model.metabolites.get_by_id(\"Diphosphate[c]\")] = -nt\n",
    "\n",
    "model.reactions.BRNA.subtract_metabolites(original)\n",
    "model.reactions.BRNA.add_metabolites(m2)\n",
    "\n",
    "print(\"AFTER:\")\n",
    "print(f\"BIOMASS = {model.slim_optimize()}\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Show the new DNA/RNA reactions:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "BDNA: 22.44 ATP[c] + 22.44 H2O[c] + 1.42514 dATP[c] + 0.64218 dCTP[c] + 0.64218 dGTP[c] + 1.42514 dTTP[c] --> 22.44 ADP[c] + 2.06732 Diphosphate[c] + 22.44 Orthophosphate[c] + 2.0 deoxyribonucleic_acids[c]\n",
      "BRNA: 8.23 ATP[c] + 0.61 CTP[c] + 0.89 GTP[c] + 7.44 H2O[c] + 0.81 UTP[c] --> 7.44 ADP[c] + 3.1 Diphosphate[c] + 7.44 Orthophosphate[c] + ribonucleic_acids[c]\n"
     ]
    }
   ],
   "source": [
    "print(model.reactions.BDNA)\n",
    "print(model.reactions.BRNA)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Generate the Virus biomass reaction\n",
    "\n",
    "We print the resulting biomass stoichiomtery, and also output the range of optimal biomass. Note that optimal phage growth precludes host growth.\n",
    "\n",
    "The resulting model is written to file, with host biomass as the default objective."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      " \n",
      "+20.446 x ADP[c]\n",
      "+00.009 x AMP[c]\n",
      "-20.455 x ATP[c]\n",
      "+00.005 x CMP[c]\n",
      "-00.005 x CTP[c]\n",
      "+01.249 x Diphosphate[c]\n",
      "+00.006 x GMP[c]\n",
      "-00.006 x GTP[c]\n",
      "-00.422 x Glycine[c]\n",
      "-20.446 x H2O[c]\n",
      "+20.446 x H[c]\n",
      "-00.428 x L_Alanine[c]\n",
      "-00.195 x L_Arginine[c]\n",
      "-00.312 x L_Asparagine[c]\n",
      "-00.300 x L_Aspartate[c]\n",
      "-00.023 x L_Cysteine[c]\n",
      "-00.247 x L_Glutamate[c]\n",
      "-00.180 x L_Glutamine[c]\n",
      "-00.050 x L_Histidine[c]\n",
      "-00.269 x L_Isoleucine[c]\n",
      "-00.310 x L_Leucine[c]\n",
      "-00.192 x L_Lysine[c]\n",
      "-00.090 x L_Methionine[c]\n",
      "-00.185 x L_Phenylalanine[c]\n",
      "-00.164 x L_Proline[c]\n",
      "-00.389 x L_Serine[c]\n",
      "-00.451 x L_Threonine[c]\n",
      "-00.039 x L_Tryptophan[c]\n",
      "-00.196 x L_Tyrosine[c]\n",
      "-00.321 x L_Valine[c]\n",
      "+20.446 x Orthophosphate[c]\n",
      "+00.007 x UMP[c]\n",
      "-00.007 x UTP[c]\n",
      "-00.378 x dATP[c]\n",
      "-00.233 x dCTP[c]\n",
      "-00.233 x dGTP[c]\n",
      "-00.378 x dTTP[c]\n",
      "                  minimum       maximum\n",
      "PHM2_prodrxn_VN  0.150551  1.505510e-01\n",
      "BIOMASS          0.000000 -1.679873e-14\n"
     ]
    }
   ],
   "source": [
    "generate_virus_biomass(virus_record=virus_record, model=model)\n",
    "print(\" \")\n",
    "for m, s in sorted(\n",
    "    model.reactions.PHM2_prodrxn_VN.metabolites.items(), key=lambda x: x[0].id\n",
    "):\n",
    "    print(f\"{s:+07.3f} x {m.id}\")\n",
    "model.objective = model.reactions.PHM2_prodrxn_VN\n",
    "fva = cobra.flux_analysis.flux_variability_analysis(\n",
    "    model, [model.reactions.PHM2_prodrxn_VN, model.reactions.BIOMASS]\n",
    ")\n",
    "print(fva)\n",
    "\n",
    "# reset the objective before writing to file\n",
    "model.objective = model.reactions.BIOMASS\n",
    "\n",
    "write_sbml_model(model, \"model_files/ProchlorococcusMED4+PHM2v1.xml\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Explore mass composition\n",
    "\n",
    "We now examine the DNA/Protein mass ratio for the phage, as well as the total mass of other (unpacked) components required for phage assembly, such as ATP."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "dna_mass=0.3992977448914946, protein_mass=0.6007022551085055, rna_mass=0.013223616993168272\n"
     ]
    }
   ],
   "source": [
    "# reverse the id strings\n",
    "met_id_lookup = {v: k for k, v in MET_IDS.items()}\n",
    "\n",
    "dna_mass = 0\n",
    "protein_mass = 0\n",
    "rna_mass = 0\n",
    "for m, s in sorted(\n",
    "    model.reactions.PHM2_prodrxn_VN.metabolites.items(), key=lambda x: x[0].id\n",
    "):\n",
    "    if s > 0:\n",
    "        continue  # only looking at consumption right now\n",
    "    if m.id in [\"dATP[c]\", \"dCTP[c]\", \"dGTP[c]\", \"dTTP[c]\"]:\n",
    "        mp = m.id[0:2] + \"M\" + m.id[3:]\n",
    "        lookup = met_id_lookup[mp]\n",
    "        dna_mass -= NUCLEOTIDE_MASS[lookup] * s / 1000\n",
    "    elif m.id in [\"ATP[c]\", \"CTP[c]\", \"GTP[c]\", \"UTP[c]\"]:\n",
    "        lookup = met_id_lookup[m.id]\n",
    "        if m.id == \"ATP[c]\":  # for ATP, we separate out energetic and RNA requirements\n",
    "            sd = model.reactions.PHM2_prodrxn_VN.metabolites[\n",
    "                model.metabolites.get_by_id(\"H2O[c]\")\n",
    "            ]\n",
    "            rna_mass -= NUCLEOTIDE_MASS[lookup] * (s - sd) / 1000\n",
    "            # other_masses -= NUCLEOTIDE_MASS[lookup] * sd / 1000\n",
    "        else:\n",
    "            rna_mass -= NUCLEOTIDE_MASS[lookup] * s / 1000\n",
    "    elif m.id.startswith(\"L_\") or m.id == \"Glycine[c]\":\n",
    "        lookup = met_id_lookup[m.id]\n",
    "        protein_mass -= AMINO_ACID_MASS[lookup] * s / 1000\n",
    "\n",
    "print(f\"{dna_mass=}, {protein_mass=}, {rna_mass=}\")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": ".venv",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
